src/odemis/util/spot.py

```diff
@@ -126,86 +128,260 @@ def SpotIntensity(data, background=None):
+
+    # thresholding
+    qradius = max(nsize // 4, 1)
+    BWdil = ndimage.binary_dilation(binary, structure=sedisk(qradius))
+
+    labels, num_features = ndimage.label(BWdil)
+    index = np.arange(1, num_features + 1)
+    mean = ndimage.mean(filtered, labels=labels, index=index)
+    si = np.argsort(mean)[::-1][:qty]
+    # estimate spot center position
+    pos = np.array(ndimage.center_of_mass(filtered, labels=labels,
+                                          index=index[si]))[:, ::-1]
+    if np.any(np.isnan(pos)):
+        logging.warning('not all positions found')
+        pos = pos[np.any(~np.isnan(pos), axis=1)]
+    # improve center estimate using radial symmetry method
+    w = nsize // 2
+    refined_position = np.zeros_like(pos)
+    for idx, xy in enumerate(pos):
+        _ij = np.rint(xy).astype(int)
+        i0, j0 = _ij - w + 1
+        i1, j1 = _ij + w
+        # if spot is close to edge of image i0 and j0 can be smaller than 0, crop so the spot is still in the center.
+        if i0 < 0:
+            i1, j1 = np.array([i1, j1]) + i0
+            i0, j0 = np.array([i0, j0]) - i0
+        if j0 < 0:
+            i1, j1 = np.array([i1, j1]) + j0
+            i0, j0 = np.array([i0, j0]) - j0
+        spot = filtered[j0:j1, i0:i1]
+        refined_position[idx] = np.rint(xy) + np.array(FindCenterCoordinates(spot))
+
+    return refined_position
+
+
+def EstimateLatticeConstant(pos):
+    """
+    Estimate the lattice constant of a point set that represent a square grid.
+
+    Parameters
+    ----------
+    pos : array like
+        A 2D array of shape (N, 2) containing the coordinates of the points.
+
+    Returns
+    -------
+    kxy : lattice constants
+    """
+    # Find the closest 4 neighbours (excluding itself) for each point.
+    tree = KDTree(pos)
+    dd, ii = tree.query(pos, k=5)
+    dr = dd[:, 1:]
+
+    # Determine the median radial distance and filter all points beyond
+    # 2*sigma.
+    med = np.median(dr)
+    std = np.std(dr)
+    outliers = np.abs(dr - med) > 2. * std  # doesn't work well if std is very high
+
+    # Determine horizontal and vertical distance (only radial distance is
+    # returned by tree.query).
+    dpos = pos[ii[:, 0, np.newaxis]] - pos[ii[:, 1:]]
+    dx, dy = dpos[:, :, 0], dpos[:, :, 1]
+    assert np.all(np.abs(dr - np.hypot(dx, dy)) < 1.0e-12)
+    # Use k-means to group the points into two directions.
+    X = np.column_stack((dx[~outliers], dy[~outliers]))
+    X[X[:, 0] < -0.5 * med] *= -1.
+    X[X[:, 1] < -0.5 * med] *= -1.
+
+    centroids, _ = kmeans(X, 2)
+    tree = KDTree(centroids)
+    _, labels = tree.query(X)
+    kxy = np.vstack((np.median(X[labels.ravel() == 0], axis=0),
+                     np.median(X[labels.ravel() == 1], axis=0)))
+
+    # The angle between the two directions should be close to 90 degrees.
+    alpha = np.math.atan2(np.linalg.norm(np.cross(*kxy)), np.dot(*kxy))
+    if np.abs(alpha - np.pi / 2.) > np.deg2rad(2.5):
```

Comment:
If all these variables are scalar, I'd suggest to use the standard python functions (`abs`, `math.pi`, `math.radians`). It'd make it more obvious about the type of all the variables.